api/v1beta1/machine_webhook.go

```diff
@@ -60,6 +61,10 @@ func (m *Machine) Default() {
 		normalizedVersion := "v" + *m.Spec.Version
 		m.Spec.Version = &normalizedVersion
 	}
+
+	if m.Spec.NodeDeletionTimeout.Duration == 0 {
+		m.Spec.NodeDeletionTimeout.Duration = 10 * time.Millisecond
```

Comment:
Should this be 
```suggestion
		m.Spec.NodeDeletionTimeout.Duration = 10 * time.Second
```
Also, what about adding a const for this value?

Reply:
@vincepri 
> * Have `NodeDeletionTimeout` to be calculated from the Kubernetes Node `DeletionTimestamp` field, if the Kubernetes Node cannot be reached, use the Machine's one instead.

Does that mean you want to wait until the node got removed from the cluster? Currently I only wait until the deletion is successful, which effectively only sets the deletion timestamp. Since nodes don't have a finalizer (at least by default), I'm not sure if it's worth to wait until it is deleted.

> * The controller should then:
>   
>   * If the `NodeDeletionTimeout` is set to `0`, always wait indefinitely and never move on and remove the finalizer.
>   * If the `NodeDeletionTimeout` is greater than `0`, calculate if the timeout has expired from the `DeletionTimestamp`. If the timeout has been reached, move on and remove the finalizer; otherwise return an error and requeue the reconciliation.

I like this approach as it makes the timeout more logical from a capi perspective. I would attempt deletion (with it's 10s timeout) before checking if the `NodeDeletionTimeout` has run out, so in case something else takes a long time we try node deletion at least once. That would also allow to replicate the previous behaviour by setting a very short timeout.

I would default the `NodeDeletionTimeout` to `0` then.